src/Workspaces/Core/Portable/Workspace/Solution/CachedSkeletonReferences.cs

```diff
@@ -4,246 +4,407 @@
         {
-            // whoever won, still store this reference set against us with the provided version.
             _version = version;
-            _skeletonReferenceSet = referenceSet;
+            _skeletonReferenceSet = skeletonReferenceSet;
         }
 
-        return referenceSet.GetMetadataReference(properties);
-    }
-
-    private SkeletonReferenceSet GetOrBuildReferenceSet(
-        Workspace workspace,
-        VersionStamp version,
-        Compilation finalCompilation,
-        CancellationToken cancellationToken)
-    {
-        // First see if we already have a reference set for this version.  if so, we're done and can return that.
-        var referenceSet = TryGetReferenceSet(version);
-        if (referenceSet != null)
+        /// <summary>
+        /// Produces a copy of the <see cref="CachedSkeletonReferences"/>, allowing forks of <see cref="ProjectState"/> to
+        /// reuse <see cref="MetadataReference"/>s when their dependent semantic version matches ours.  In the case where
+        /// the version is different, then the clone will attempt to make a new skeleton reference for that version.  If it
+        /// succeeds, it will use that.  If it fails however, it can still use our skeletons.
+        /// </summary>
+        public CachedSkeletonReferences Clone()
         {
-            workspace.LogTestMessage($"Succeeded at finding reference set corresponding to requested version.");
-            return referenceSet;
+            lock (_stateGate)
+            {
+                // pass along the best version/reference-set we computed for ourselves.  That way future ProjectStates
+                // can use this data if either the version changed, or they weren't able to build a skeleton for themselves.
+                // By passing along a copy we ensure that if they have a different version, they'll end up producing a new
+                // SkeletonReferenceSet where they'll store their own data in which will not affect prior ProjectStates.
+                return new CachedSkeletonReferences(_version, _skeletonReferenceSet);
+            }
         }
 
-        // okay, we don't have one. so create one now.
+        public async Task<MetadataReference?> GetOrBuildReferenceAsync(
+            ICompilationTracker compilationTracker,
+            SolutionState solution,
+            MetadataReferenceProperties properties,
+            CancellationToken cancellationToken)
+        {
+            // First, just see if we have cached a reference set that is complimentary with the version of the project
+            // being passed in.  If so, we can just reuse what we already computed before.
+            var workspace = solution.Workspace;
+            var version = await compilationTracker.GetDependentSemanticVersionAsync(solution, cancellationToken).ConfigureAwait(false);
+            var metadataReference = TryGetReferenceSet(version)?.GetMetadataReference(properties);
```

Comment:
the caching that is going on here is a bit confusing. 
1.  we first look at the `_skeletonReferenceSet` from `TryGetReferenceSet`
2.  Assuming that isn't present, it goes to `GetOrBuildReferenceSetAsync` which immediately looks at the `s_compilationToReferenceMap` (in `TryGetExistingReferenceSet`) for the compilation with a note mentioning that we want to return the same instance for a compilation.  Should 2) be done before 1)?   Or is 2) done after to avoid looking at the compilation if not needed?
3.  `TryGetExistingReferenceSet` then calls `TryGetReferenceSet`, but we already looked at that in 1).  Something doesn't add up

To me it is quite unclear which cache is preferred and when an item should be in `s_compilationToReferenceMap` vs. `_skeletonReferenceSet`